src/Stack/BuildPlan.hs

```diff
@@ -667,3 +678,43 @@ shadowMiniBuildPlan (MiniBuildPlan ghc pkgs0) shadowed =
                 Just True -> Left
                 Just False -> Right
                 Nothing -> assert False Right
+
+parseCustomMiniBuildPlan :: (MonadIO m, MonadCatch m, MonadLogger m, MonadReader env m, HasHttpManager env, HasConfig env, MonadBaseControl IO m)
+                         => T.Text -> m MiniBuildPlan
+parseCustomMiniBuildPlan url = do
+    root <- asks $ configStackRoot . getConfig
+    let hashStr = S8.unpack $ B16.encode $ SHA256.hash $ encodeUtf8 url
+    hashFP <- parseRelFile hashStr
+    hashFPBin <- parseRelFile $ hashStr ++ ".bin"
+    let cacheDir = root </> $(mkRelDir "custom-plan-cache")
+        binaryFP = cacheDir </> hashFPBin
+    taggedDecodeOrLoad (toFilePath binaryFP) $ do
+        fp <-
+            case parseUrl $ T.unpack url of
+                Just req -> do
+                    let cacheFP = cacheDir </> hashFP
+                    _ <- download req cacheFP
+                    return cacheFP
+                Nothing -> do
+                    fp <- liftIO $ canonicalizePath $ T.unpack $ fromMaybe url $
+                        T.stripPrefix "file://" url <|> T.stripPrefix "file:" url
+                    parseAbsFile fp
+        cs <- liftIO (Data.Yaml.decodeFileEither $ toFilePath fp) >>= either throwM return
+        let addFlags :: PackageIdentifier -> (PackageName, (Version, Map FlagName Bool))
+            addFlags (PackageIdentifier name ver) =
+                (name, (ver, fromMaybe Map.empty $ Map.lookup name $ csFlags cs))
+        toMiniBuildPlan
+            (csGhcVersion cs)
+            Map.empty
+            (Map.fromList $ map addFlags $ Set.toList $ csPackages cs)
+
+data CustomSnapshot = CustomSnapshot
+    { csGhcVersion :: !Version
+    , csPackages :: !(Set PackageIdentifier)
+    , csFlags :: !(Map PackageName (Map FlagName Bool))
+    }
+instance FromJSON CustomSnapshot where
+    parseJSON = withObject "CustomSnapshot" $ \o -> CustomSnapshot
+        <$> o .: "ghc"
+        <*> o .: "packages"
+        <*> o .:? "flags" .!= Map.empty
```

Comment:
Certainly not necessary at this point, but it could be quite nice to have a "base" on which this snapshot builds, similar to how `extra-deps` overrides the snapshot. You could imagine a hierarchy of snapshots which cascade into the final product. I would use it to base a snapshot on one of stackage's snapshots, and override a few things.

This seems like it would be a lot more work for not a lot more gain, but I figured I'd mention it.